### ISSUE TITLE
web3.eth.getCompilers and web3.eth.compile.solidity fail on mainnet, after most recent update

### ISSUE THREAD

**Issue by Tectract** [MetaMask/metamask-plugin#1493]
In reference to this issue: https://github.com/MetaMask/metamask-plugin/issues/1490

Both these functions work ok on localhost:8545 against testRPC, but fail on mainnet after the most recent update:

web3.eth.getCompilers(function(err,resp){ if(err){ console.log("err: " + err); } console.log("resp: " + resp); })

web3.eth.compile.solidity("pragma stuff...",function(err, resp){ if(err){ console.log("err: " + err); } console.log("resp: " + resp); })

**Comment by flyswatter** [comment]
We actually haven't supported any of the compiler methods before, I believe they're being deprecated from the providers anyway. This is why Remix brings its own VM & compilers.

I guess we could reword this as a request to add compiler support into MetaMask?

**Comment by flyswatter** [comment]
Also see:
https://github.com/ethereum/go-ethereum/issues/3793

**Comment by flyswatter** [comment]
Also: https://github.com/ethereum/EIPs/issues/209

Apparently this has been cooking for a while. It's too bad deprecation warnings haven't become a more prominent practice in this ecosystem.

**Comment by Tectract** [comment]
Well, the thing is that it was working perfectly fine until this last update.

No longer being able to use metamask to compile and deploy to the mainnet is a huge depracation of functionality, for a lot of users. We were all really excited that we could FINALLY deploy smart-contracts to the mainnet without needing an actual full node running on our own. 

Is there some way to test against the most recent revision before this one? Some dependency slipped in the last 20 days...

Please see :

http://www.enledger.io/EthDeployer
https://github.com/Tectract/EthDeployer

**Comment by flyswatter** [comment]
I'm really sorry, like I tried to demonstrate with the links I provided, this is not in our hands at all. We forward to ethereum clients.  The client developers decided to remove this functionality.  I think breaking APIs is a really awful practice, but they did it, so if you'd like to see the methods returned, you should express it to them.

I made [a comment on the original EIP](https://github.com/ethereum/EIPs/issues/209#issuecomment-303881823) to help express the issue that was created by this deprecation. I'm not sure the client devs are being fully considerate of dapp developers like yourself.

**Comment by Tectract** [comment]
re: https://github.com/ethereum/go-ethereum/issues/3793

I'm not sure I understand where this dependency is in the pipeline. If it's server-side I could just rollback, but I don't think anything changed on the enledger.io webserver recently...

Strange, one of the workarounds listed in that thread is:

To generate bytecode:
> solc --bin soliditySource.sol

But... that's what it was doing before. I had to specifically set up solc on the webserver and make it available to apache user to enable the EthDeployer to work, in the first place.

**Comment by Tectract** [comment]
Can you recommend any possible workaround?

**Comment by flyswatter** [comment]
You could run an older Ethereum client, and point your MetaMask instance at it.

**Comment by flyswatter** [comment]
Alternatively, you could bring a JS compiler in with your website, and compile within the site itself.

**Comment by flyswatter** [comment]
Our hosted providers (that we point to by default) are running the latest clients provided by the Geth and Parity teams, and when they remove a feature, our users are affected.

Since we forward those methods to our full-node backends, we don't have control over those feature removals.

**Comment by Tectract** [comment]
I'm not running any ethereum client on the webserver, that was the whole point of using metamask. It WAS the connection to the mainnet. I have solc installed on the webserver and available to the apache user, and that was enough for everything to be happy, two weeks ago. None of my client-side or server-side code has changed. 

So it's on the servers that you guys are running, I think, where the deprecation has occured. It must be the fact that those function calls have been removed from your instances of Geth. I would highly recommend rolling back your version of Geth until this issue is resolved, because MetaMask just lost one of it's most important and cool features: the ability to enable users to compile and deploy to mainnet without running their own full node.

**Comment by Tectract** [comment]
> Alternatively, you could bring a JS compiler in with your website, and compile within the site itself.

This seems like it may be my best option. But I'm not sure I understand it.

Can you elaborate a little more on this? Is there some way to use Remix or something to compile the contract within the browser, client-side?

**Comment by Tectract** [comment]
Maybe I could make an end-point on my webserver that literally just tries to compile stuff via solc --bin, and then just AJAX against it... hmmmmm

**Comment by Tectract** [comment]
The in-browser work-around is highly more desireable, so that the EthDeployer can stay as a simple react app without needing advanced server customizations...

**Comment by Tectract** [comment]
This looks relevant for compiling solc code client-side, I'll give it a shot!

http://hypernephelist.com/2017/01/19/deploy-ethereum-smart-contract-using-client-signature.html

**Comment by axic** [comment]
Note, you can use the Solidity compiler in your Dapp directly, see https://github.com/ethereum/solc-js for instructions.

**Comment by flyswatter** [comment]
Here is the solc client side library:
https://www.npmjs.com/package/solc

**Comment by flyswatter** [comment]
Oop, Axic beat me to it.

**Comment by kumavis** [comment]
Here is an example of using the solc module
I believe this is how testrpc does it
Should work in browser without modification
https://github.com/MetaMask/provider-engine/blob/master/subproviders/solc.js

**Comment by Tectract** [comment]
I will update this tonight or in the morning. This actually makes the EthDeployer much nicer because it means you no longer need special install of solc in the web-server :)

Thanks guys! Will post back when the BlockGeeks "Intro to Solidity and Smart-Contracts" article drops, Metamask will be featured! Cheers~

**Comment by Tectract** [comment]
Ugh, still having a really hard time with this. 

Since EthDeployer is a react app, I cannot just do:

```

**Comment by Tectract** [comment]
let me try this...

https://github.com/ericxtang/browser-solc

**Comment by Tectract** [comment]
Argh, can't get any of this working. Nodejs solc module does not have a commonJS entry point... browser-solc is not happy getting minified and included for some reason. hmmmmmmmmmmmmmmmmm
